File: src/parser.cpp
#include "lexer.hpp"
#include "parser.hpp"
#include <string>
#include <iostream>
#include <vector>

using std::string;

Parser::Parser(std::vector<Token> _tokens) {
    tokens = _tokens;
}

Token Parser::advance() {
    TokenIndex += 1;
    if (TokenIndex < tokens.size()) {
        CurrentTok = tokens[TokenIndex];
    }
    return CurrentTok;
}

Node Parser::factor() {
    Token token = CurrentTok;

    if (token.type == "integer" || token.type == "float") {
        std::cout << "Test\n";
        advance();
        return Node(token);
    }
}

Node Parser::term() {
    Node left = factor();

    while (is_operator(CurrentTok.value) == "multiply" || is_operator(CurrentTok.value) == "divide") {
        Token OperatorToken = CurrentTok;
        advance();
        Node right = factor();
        left = Node(OperatorToken, &left, &right);
    }

    return left;
}

Node Parser::expression() {
    Node left = term();

    while (is_operator(CurrentTok.value) == "add" || is_operator(CurrentTok.value) == "subtract") {
        Token OperatorToken = CurrentTok;
        advance();
        Node right = term();
        left = Node(OperatorToken, &left, &right);
    }

    return left;
}

Node Parser::Parse() {
    Node result = expression();
    return result;
}

/*
void parse(std::vector<Token> tokenVec, int length) {
    std::vector<Node> syntaxTree;
    for (int i = 0; i < length; i++) {
        if (is_operator(tokenVec[i].value) != "") {
            Node node = {tokenVec[i].value, tokenVec[i].type, new Node({tokenVec[i-1].value, tokenVec[i-1].type, NULL, NULL}), new Node({tokenVec[i+1].value, tokenVec[i+1].type, NULL, NULL})};
            for (int j = 0; j < length; j++) {
                std::cout << node.right->data << "\n";
                if (is_operator(tokenVec[j].value) != "") {
                    node.right = new Node({tokenVec[j].value, tokenVec[j].type, new Node({tokenVec[j-1].value, tokenVec[j-1].type, NULL, NULL}), new Node({tokenVec[j+1].value, tokenVec[j+1].type, NULL, NULL})});
                }
            }
            syntaxTree.push_back(node);
        }
    }

    for (int i = 0; i < syntaxTree.size(); i++) {
        std::cout << syntaxTree[i].left->data << " -> " << syntaxTree[i].data << " <- " << syntaxTree[i].right->data << "\n";
    }
}
*/

File: typescript/parser.ts
import { Token } from './lexer'

export class NumberNode {
    _value: string;
    _type: string;

    constructor(token: Token) {
        this._value = token.value;
        this._type = token.type;
    }

    get value(): string {
        return this._value;
    }

    setValue(newValue: string): void {
        this._value = newValue;
    }

    get type(): string {
        return this._type;
    }

    setType(newType: string): void {
        this._type = newType;
    }
}

export class BinOpNode {
    private _value: string;
    private _type: string;
    private _left;
    private _right;

    constructor(token: Token, left, right) {
        this._value = token.value;
        this._type = token.type;
        this._left = left;
        this._right = right;
    }

    get value(): string {
        return this._value;
    }

    setValue(newValue: string): void {
        this._value = newValue;
    }

    get type(): string {
        return this._type;
    }

    setType(newType: string): void {
        this._type = newType;
    }

    get left() {
        return this._left;
    }

    setLeft(node): void {
        this._left = node;
    }

    get right() {
        return this._right;
    }

    setRight(node): void {
        this._right = node;
    }
}

export class Parser {
    public tokens: Token[];
    public tokenIdx: number = -1;
    public currentTok: Token;

    // constructor
    constructor(tokens: Token[]) {
        this.tokens = tokens;
        this.advance();
    }

    // advance
    advance(): Token {
        this.tokenIdx += 1;
        if (this.tokenIdx < this.tokens.length) {
            this.currentTok = this.tokens[this.tokenIdx];
        }
        return this.currentTok;
    }

    // factor
    factor() {
        let token: Token = this.currentTok;

        if (token.type == "integer" || token.type == "float") {
            this.advance();
            return new NumberNode(token);
        }
    }

    // term
    term() {
        let left: any = this.factor();

        while (this.currentTok.type == "multiply" || this.currentTok.type == "divide") {
            let OperatorToken: Token = this.currentTok;
            this.advance();
            let right = this.factor();
            left = new BinOpNode(OperatorToken, left, right);
        }

        return left;
    }

    // expression
    expression() {
        let left: any = this.term();
        
        while (this.currentTok.type == "add" || this.currentTok.type == "subtract") {
            let OperatorToken: Token = this.currentTok;
            this.advance();
            let right = this.term();
            left = new BinOpNode(OperatorToken, left, right);
        }
        return left;
    }

    // Parse
    Parse() {
        let result: BinOpNode = this.expression();
        return result;
    }
}